src/Control/Monad/Eff/AVar.js

```diff
@@ -0,0 +1,311 @@
+    queue.last.prev = null;
+    queue.head = queue.last;
+    queue.last = null;
+    break;
+  default:
+    cell = queue.head;
+    queue.head = cell.next;
+    queue.head.prev = null;
+  }
+  cell.next = null;
+  cell.queue = null;
+  queue.size--;
+  return cell.value;
+}
+
+function deleteCell (cell) {
+  if (cell.queue === null) {
+    return;
+  }
+  if (cell.queue.last === cell) {
+    takeLast(cell.queue);
+    return;
+  }
+  if (cell.queue.head === cell) {
+    takeHead(cell.queue);
+    return;
+  }
+  if (cell.prev) {
+    cell.prev.next = cell.next;
+  }
+  if (cell.next) {
+    cell.next.prev = cell.prev;
+  }
+  cell.queue.size--;
+  cell.queue = null;
+  cell.value = null;
+  cell.next  = null;
+  cell.prev  = null;
+}
+
+function AVar () {
+  this.draining = false;
+  this.error    = null;
+  this.value    = EMPTY;
+  this.takes    = new MutableQueue();
+  this.reads    = new MutableQueue();
+  this.puts     = new MutableQueue();
+}
+
+exports.makeEmptyVar = function () {
+  return new AVar();
+};
+
+exports.makeVar = function (value) {
+  return function () {
+    var avar = new AVar();
+    avar.value = value;
+    return avar;
+  };
+};
+
+exports._killVar = function (left, right, avar, error) {
+  return function () {
+    if (avar.error === null) {
+      avar.error = error;
+      avar.value = EMPTY;
+      drainVar(left, right, avar);
+    }
+  };
+};
+
+exports._putVar = function (left, right, avar, value, cb) {
+  return function () {
+    if (avar.error !== null) {
+      runEff(cb(left(avar.error)));
+      return NO_EFFECT;
+    }
+    var cell = putLast(avar.puts, { cb: cb, value: value });
+    drainVar(left, right, avar);
+    return function () {
+      deleteCell(cell);
+    };
+  };
+};
+
+exports._takeVar = function (left, right, avar, cb) {
+  return function () {
+    if (avar.error !== null) {
+      runEff(cb(left(avar.error)));
+      return NO_EFFECT;
+    }
+    var cell = putLast(avar.takes, cb);
+    drainVar(left, right, avar);
+    return function () {
+      deleteCell(cell);
+    };
+  };
+};
+
+exports._readVar = function (left, right, avar, cb) {
+  return function () {
+    if (avar.error !== null) {
+      runEff(cb(left(avar.error)));
+      return NO_EFFECT;
+    }
+    var cell = putLast(avar.reads, cb);
+    drainVar(left, right, avar);
+    return function () {
+      deleteCell(cell);
+    };
+  };
+};
+
+exports._tryPutVar = function (left, right, avar, value) {
+  return function () {
+    if (avar.value === EMPTY && avar.error === null) {
+      avar.value = value;
+      drainVar(left, right, avar);
+      return true;
+    } else {
+      return false;
+    }
+  };
+};
+
+exports._tryTakeVar = function (left, right, nothing, just, avar) {
+  return function () {
+    var value = avar.value;
+    if (value === EMPTY) {
+      return nothing;
+    } else {
+      avar.value = EMPTY;
+      drainVar(left, right, avar);
+      return just(value);
+    }
+  };
+};
```

Comment:
Do we not need to check `error` in this case? I see it's handled in `drainVar` too, but just wanted to double check, as it's checked in the non-try `takeVar`.